Drive rotation commands

lilbites21 opened the pull request:
added 2 commands for later use to rotate the robot and to set its angle.

lilbites21 committed:
Added command to rotate to set degree

lilbites21 committed:
Made a rotate a set amount command

lilbites21 committed:
Deleted unnecessary variables

msoucy commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+        };
+    }
+
+    public CommandBase setRelitiveAngle(final Rotation2d angle, final double speed) {
```
Not how "relative" is spelled

bot190 replied:
```suggestion
    public CommandBase setRelativeAngle(final Rotation2d angle, final double speed) {
```

msoucy commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+        return new CommandBase() {
+            {
+                addRequirements(thisDrive);
+                setName("Relitive Angle");
```
Still not how "relative" is spelled

msoucy commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+            @Override
+            public boolean isFinished() {
+                return Math
+                        .abs(pose.getRotation().times(1).getDegrees() - targetRotation.getDegrees()) <= ROTATION_BUFFER;
```
Why use `.times(1)` when you're just doing math on it?

More importantly, if you overshoot what will happen?

msoucy commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+
+            @Override
+            public void initialize() {
+                targetRotation = new Rotation2d(pose.getRotation().times(1).getDegrees() + angle.getDegrees());
```
If you're just getting a value from the rotation you don't need `.times(1)` here.

msoucy commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+                setName("Relitive Angle");
+            }
+            private double speed2 = speed;
+            private Rotation2d targetRotation;
```
Since this entire variable is set once, why not create it with `thisDrive` up above?

msoucy commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+            @Override
+            public void initialize() {
+                targetRotation = new Rotation2d(pose.getRotation().times(1).getDegrees() + angle.getDegrees());
+                speed2 *= Math.signum(angle.getDegrees());
```
This math can be done up above with `thisDrive` instead of as a member.

bot190 commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+
+            @Override
+            public boolean isFinished() {
+                return Math.abs(pose.getRotation().times(1).getDegrees() - angle.getDegrees()) <= ROTATION_BUFFER;
```
What happens if we were to rotate past the desired angle by accident (rotating to fast)?

bot190 commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
+                addRequirements(thisDrive);
+                setName("Absolute angle");
+            }
+            private double speed2 = speed;
```
I would prefer to change the argument name to `aSpeed` instead of making this `speed2`

msoucy commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -146,6 +147,77 @@ public void end(boolean interrupted) {
         };
     }
 
+    public CommandBase setAbsoluteAngle(final Rotation2d angle, final double speed) {
```
All comments on `setRelativeAngle` also apply here.

Are the functions similar enough that you can extract common logic?

bot190 reviewed (changes requested):
We probably want to use a PID controller to get us to a given angle long term. But for now we can probably get away with a bang-bang controller. That is, rotate towards the target but if we overshoot by too much we need to re-evaluate what direction we need to rotate. Maybe each time we do this we reduce our rotational speed so we're less likely to rotate past the target.

lilbites21 committed:
Delete unnecessary lines, moved variable assignment

lilbites21 committed:
My attempt at stopping over shoot.

I think it works, but if it doesn't... It's about the journey not the destination.

lilbites21 committed:
Moved initializations of variable back into initialize method

there was a misunderstanding

lilbites21 committed:
Moved PersistenceCheck to the correct spot

msoucy commented:
Deprecated in favor of the autonomous-test branch
